### PR TITLE
Activities re-write, part 2 - Child and composite activities

### PR DESCRIPTION
**Pull request by obrakmann**
I'm confident enough with the code now to inflict it on reviewers to shake out the last bugs.

### CODE REVIEW

File: OpenRA.Game/Activities/Activity.cs [hunk 1]
```diff
@@ -9,50 +9,218 @@
...
+				ChildActivity = activity;
+		}
+
+		protected void PrintActivityTree(Activity origin = null, int level = 0)
+		{
+			if (origin == null)
+				Root.PrintActivityTree(this);
+			else
+			{
+				Console.Write("".PadLeft(level * 2));
```
Comment (Phrohdoh) [inline OpenRA.Game/Activities/Activity.cs]: Is there any reason you went with `PadLeft` over `new string(' ', level * 2)`?

File: OpenRA.Game/Activities/Activity.cs [hunk 2]
```diff
@@ -9,50 +9,218 @@
...
+		public ActivityState State { get; private set; }
+
+		public Activity Root
+		{
+			get
+			{
+				if (Parent == null)
+					return this;
+
+				var parent = Parent;
```
Comment (Phrohdoh) [inline OpenRA.Game/Activities/Activity.cs]: What is the purpose of using `Parent` above, then shadowing the private `parent` member here?

File: OpenRA.Game/Activities/Activity.cs (hunk 1 above)
Comment (obrakmann) [inline OpenRA.Game/Activities/Activity.cs]: No reason, except I didn't know that this ctor existed, so I went with the first suggestion Google threw at me. Will change, as that is exactly what I was looking for.

File: OpenRA.Mods.Common/Activities/Parachute.cs [hunk 3]
```diff
@@ -83,6 +83,7 @@ public override Activity Tick(Actor self)
 		// Only the last queued activity (given order) is kept
 		public override void Queue(Activity activity)
 		{
+			NextActivity = null;
 			NextActivity = activity;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Parachute.cs]: This looks weird, why is this line needed?

Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Parachute.cs]: Oh, that's a left-over from when I was toying with the idea of having an assignment to `NextActivity` not override it but queue the newly assigned activity after the old one.

Fixed.

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs [hunk 4]
```diff
@@ -16,59 +16,40 @@
...
 		}
 
-		public override bool Cancel(Actor self)
+		public override Activity Tick(Actor self)
 		{
-			if (!IsCanceled && inner != null && !inner.Cancel(self))
-				return false;
-
-			return base.Cancel(self);
+			return NextActivity;
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: We definitely need an obvious comment, or maybe a rethink of `CompositeActivity` here.

My first assumption from reading this class is that `OnFirstRun` is queuing up the child activities that are then ticked by the actor code in preference to this one.  Only once they have all completed does this parent activity Tick, and, with nothing left to do, moves on to the next logical activity.  This assumption is completely wrong.

File: OpenRA.Mods.Common/Activities/Air/TakeOff.cs [hunk 5]
```diff
@@ -37,7 +37,7 @@ public override Activity Tick(Actor self)
 			var destination = rp != null ? rp.Location :
 				(hasHost ? self.World.Map.CellContaining(host.CenterPosition) : self.Location);
 
-			if (NextActivity == null)
+			if (nextActivity == null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/TakeOff.cs]: Again I think mixing `nextActivity` and `NextActivity` is dangerous.  Can we please keep this as `NextActivity == Parent`?

File: OpenRA.Mods.Common/Activities/DeliverResources.cs [hunk 6]
```diff
@@ -35,7 +35,7 @@ public DeliverResources(Actor self)
 
 		public override Activity Tick(Actor self)
 		{
-			if (NextActivity != null)
+			if (nextActivity != null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/DeliverResources.cs]: Here too.

File: OpenRA.Mods.Common/Activities/FindResources.cs [hunk 7]
```diff
@@ -51,7 +51,7 @@ public FindResources(Actor self, CPos avoidCell)
 
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled || NextActivity != null)
+			if (IsCanceled || nextActivity != null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: Here too.

**Review by pchote — CHANGES_REQUESTED** [review]
Looks good overall, but see the comments below.  I haven't tested this in game yet.

**Comment by pchote** [comment]
Would you mind splitting c620870d808ee8adfe5062af86b49389b51aff90 into its own PR?  That can then be trivially reviewed and merged, making it a bit easier to focus on the fiddly bits of this PR.

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs (hunk 4 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: I think the simplest option would be to change the NextActivity override with a new ChildOrNextActivity property.

File: OpenRA.Game/Activities/Activity.cs [hunk 8]
```diff
@@ -9,50 +9,216 @@
...
+		protected Activity ChildActivity
+		{
+			get
+			{
+				return childActivity != null && childActivity.State < ActivityState.Done ? childActivity : null;
+			}
+
+			set
+			{
+				if (value == this || value == Parent || value == nextActivity)
```
Comment (obrakmann) [inline OpenRA.Game/Activities/Activity.cs]: Yup, they wouldn't be caught, possibly creating a loop. Checking more of the tree is quite a difficult and probably expensive operation. You'd not only have to check the parent's parents, but the parent's next queue as well (and the parent's parent's next queue, and their parents and so on). I went with just checking for those activities that would be likely to be returned by the activities own `Tick()` method instead.

File: OpenRA.Game/Activities/Activity.cs [hunk 9]
```diff
@@ -9,50 +9,216 @@
...
+				throw new InvalidOperationException("Actor {0} attempted to tick activity {1} after it had already completed.".F(self, this.GetType()));
+
+			if (State == ActivityState.Queued)
+			{
+				OnFirstRun(self);
+				State = ActivityState.Active;
+			}
+
+			var ret = Tick(self);
+			if (ret == null || (ret != this && ret.Parent != this))
```
Comment (obrakmann) [inline OpenRA.Game/Activities/Activity.cs]: I can't think of a situation where levels could be skipped like that. The change in `ActivityUtils` should prevent it.

File: OpenRA.Game/Traits/ActivityUtils.cs [hunk 10]
```diff
@@ -44,7 +44,7 @@ public static Activity RunActivity(Actor self, Activity act)
 				else
 					start = current;
 
-				if (prev == act)
+				if (act == prev || act == prev.Parent)
```
Comment (obrakmann) [inline OpenRA.Game/Traits/ActivityUtils.cs]: Not sure where the confusion is coming from. An activity's `Tick` method shouldn't explicitly return its `Parent`, the `NextActivity` getter already handles that. Also activities should generally never assume to know what their `Parent` is, even less so its grant parent So returning `Parent.Parent` would require the activity to know about implementation details that it doesn't have any business knowing.

The goal is to eventually have activities that are as small in scope and self-contained enough that they can be mixed and matched as needed.

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs (hunk 4 above)
Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: Yup, I already thought about that, but didn't want to change the PR's code again as if it was still a work in progress. I'm was going to add `NextInQueue` property that will check against `nextActivity` only, which is what we need.

File: OpenRA.Game/Traits/ActivityUtils.cs (hunk 10 above)
Comment (pchote) [inline OpenRA.Game/Traits/ActivityUtils.cs]: My confusion really comes down to it not being clear what `NextActivity` can/will return in different cases.  I hope my other comments gives some direction on how to fix this, even if its just adding some comments for now, to be fixed more deeply (ideally with a clear split between child and next activities) in future steps.

**Comment by obrakmann** [comment]
Thanks for the review! I'll try to find some time to address your comments during the weekend.

File: OpenRA.Game/Traits/ActivityUtils.cs (hunk 10 above)
Comment (obrakmann) [inline OpenRA.Game/Traits/ActivityUtils.cs]: The reason I went that way was to avoid touching (and auditing!) all the existing activities.

You don't normally need to know what `NextActivity` will return anyway.

Also, if you insist on undoing this "uncertainty" in the future, I'm pretty sure that the `CompositeActivity` thing will stop to work, since it relies on exactly that ambiguity to work.

Comment (pchote) [inline OpenRA.Game/Traits/ActivityUtils.cs]: I might be missing something, but I currently don't see the problem.  The main thing i'm suggesting is a global replacement of NextActivity -> NextOrChildActivity in the traits being migrated to `CompositeActivity`.  I'd *really* like to know when reading the code whether an activity is finishing for good vs potentially deferring to a child activity, without having to remember to check whether it is a normal or composite activity.

Comment (obrakmann) [inline OpenRA.Game/Traits/ActivityUtils.cs]: What you are missing is that it would make both types of activities look and behave differently to the outside world. Remember that `CompositeActivity` is a kind-of `Activity`, so it needs to act like that. If you add a new property instead of overriding `NextActivity`, code like [this](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/AI/States/AirStates.cs#L136) might break unless you check what kind of activity you're dealing with. The way I wrote this, it's all transparent to users and they don't need to care at all (yes, I realize that's kind of what you're complaining about).

Comment (pchote) [inline OpenRA.Game/Traits/ActivityUtils.cs]: Ok, that makes sense.  You've got a better picture of all the moving parts than me, so I trust that this will work out by the end.  More comments though, please!

File: OpenRA.Mods.Common/Activities/FindResources.cs [hunk 11]
```diff
@@ -51,7 +51,7 @@ public FindResources(Actor self, CPos avoidCell)
 
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled || NextActivity != null)
+			if (IsCanceled || NextInQueue != null)
 				return NextActivity;
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: Is checking `NextInQueue` but returning `NextActivity` intentional?

Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: Yes. If it's canceled, `nextActivity` will be null, and we need to return to the parent.

Comment (pchote) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: Why can't we check `NextActivity != null` in this case?  Is there a situation where `NextActivity` can be null when `NextInQueue` isn't null?

Comment (pchote) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: Splitting this into two checks make this clearer?  Code that checks if something isn't null and then returns something else is usually a copy-paste error, so for my future sanity it would be nice if we could change this or have a comment `// Yes, I really do mean to check NextInQueue but return NextActivity`

Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: > Why can't we check NextActivity != null in this case?

Because `NextActivity` will be non-null as soon as there is a parent, which would then immediately quit the activity.

> Is there a situation where NextActivity can be null when NextInQueue isn't null?

No.

> Splitting this into two checks might make this clearer?

Yup, will do.

**Review by pchote — APPROVED** [review]
Looks good, just a couple of minor comments/questions (some are shown as outdated, but aren't).

File: OpenRA.Game/Activities/Activity.cs [hunk 12]
```diff
@@ -9,42 +9,256 @@
...
 	public abstract class Activity
 	{
-		public Activity NextActivity { get; set; }
+		public ActivityState State { get; private set; }
+
+		/// <summary>
+		/// Returns the top-most activity *from the point of view of the calling activity*. Note that the root activity
+		/// can and likely will have next activities of its own, which would in turn be the root for their children.
+		/// </summary>
+		public Activity RootActivity
```
Comment (obrakmann) [inline OpenRA.Game/Activities/Activity.cs]: Maybe, but it's too long for my liking.

File: OpenRA.Game/Activities/Activity.cs [hunk 13]
```diff
@@ -18,11 +18,143 @@ namespace OpenRA.Activities
...
+					next.ParentActivity = parentActivity;
+			}
+		}
+
+		Activity childActivity;
+		protected Activity ChildActivity
+		{
+			get
+			{
+				return childActivity != null && childActivity.State == ActivityState.Done ? childActivity : null;
```
Comment (obrakmann) [inline OpenRA.Game/Activities/Activity.cs]: Of course, good catch!

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs (hunk 4 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: > I'm was going to add NextInQueue property that will check against nextActivity only, which is what we need.

@obrakmann I noticed there is a `NextInQueue` now, but (still?) not used here. Is that intentional?

**Comment by reaperrr** [comment]
As far as my very limited understanding of the code goes, this looks good, and I didn't encounter any in-game regressions.

👍 once my question https://github.com/OpenRA/OpenRA/pull/12461#discussion_r98354486 has been adressed.

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs (hunk 4 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: It's intentional.  This activity explicitly relies on `NextActivity` returning the `ChildActivity`s.  `NextInQueue` is for the cases where we don't want that to happen.

Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: Ok.

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/151f005502dbce57ca989c5061664b4851cbfa34...50092dcbaa2b06c4495284c664d72d91649bdc41)
